handler.js

```diff
@@ -392,56 +402,96 @@ function discover(handler, req, head, body, cb) {
         return;
     }
 
-    var exitNodes = handler.egressNodes.exitsFor(serviceName);
-    var exitHosts = Object.keys(exitNodes);
-
-    var svcchan = null;
-    var myHost = handler.channel.hostPort;
-    if (exitHosts.indexOf(myHost) < 0 && req.headers.cn !== 'hyperbahn') {
-        // Since Hyperbahn is fully connected to service hosts,
-        // any exit node suffices.
-        svcchan = handler.channel.topChannel.handler.getOrCreateServiceChannel(serviceName);
-        handler.tchannelThrift.send(svcchan.request({
-            serviceName: 'hyperbahn',
-            headers: {
-                cn: 'hyperbahn'
-            },
-            parent: req,
-            timeout: 5000,
-            timeoutPerAttempt: 500,
-            trace: false
-        }), 'Hyperbahn::discover', null, body, function handleForward(err, resp) {
-            if (err) {
-                handler.channel.logger.error('Failed to call discover API on exit node', {
-                    error: err,
-                    serviceName: serviceName
-                });
-                return cb(err, null);
-            }
+    var topChannel = handler.channel.topChannel;
+    var svcchan = topChannel.handler.getOrCreateServiceChannel(serviceName);
+
+    if (svcchan.serviceProxyMode === 'forward' &&
+        req.headers.cn !== 'hyperbahn'
+    ) {
+        handler._forwardToRemoteDiscover(req, svcchan, body, cb);
+        return;
+    }
+
+    handler._getDiscoverHosts(serviceName, cb);
+};
 
-            cb(null, resp);
+HyperbahnHandler.prototype.discoverAffine =
+function discoverAffine(handler, req, head, body, cb) {
+    var serviceName = body.query.serviceName;
+    if (serviceName.length === 0) {
+        cb(null, {
+            ok: false,
+            body: InvalidServiceName({
+                serviceName: serviceName
+            }),
+            typeName: 'invalidServiceName'
         });
-    } else {
-        svcchan = handler.channel.topChannel.subChannels[serviceName];
-        var hosts = [];
-        if (svcchan) {
-            hosts = convertHosts(svcchan.peers.keys());
-        }
-        if (hosts.length === 0) {
-            cb(null, {
-                ok: false,
-                body: NoPeersAvailable({
-                    serviceName: serviceName
-                }),
-                typeName: 'noPeersAvailable'
-            });
-        } else {
-            cb(null, {
-                ok: true,
-                body: {
-                    peers: hosts
-                }
+        return;
+    }
+
+    handler._getDiscoverHosts(serviceName, cb);
+};
+
+HyperbahnHandler.prototype._forwardToRemoteDiscover =
+function _forwardToRemoteDiscover(parent, svcchan, body, cb) {
+    var self = this;
+    var serviceName = svcchan.serviceName;
+
+    var endpoint = self.discoveryRouteMethod;
+
+    // Since Hyperbahn is fully connected to service hosts,
```

Comment:
it's not about connected fully vs partially; it's because we fully relay all advertisements